roaring/roaring.go

```diff
@@ -395,32 +395,300 @@ func (b *Bitmap) Intersect(other *Bitmap) *Bitmap {
+				// union we can generate an RLE container that represents the entire
+				// range.
+				container := &Container{
+					runs:          []interval16{{start: 0, last: maxContainerVal}},
+					containerType: containerRun,
+					n:             maxContainerVal + 1,
+				}
+				target.Containers.Put(iKey, container)
+				bitmapIters[i:].markItersWithCurrentKeyAsHandled(iKey)
+			} else {
+				// Use a bitmap container for the target bitmap, and union everything
+				// into it.
+				//
+				// TODO(rartoul): Add another conditional case for n < ArrayMaxSize
+				// (or some fraction of that value) to avoid allocating expensive
+				// bitmaps when unioning many low-density array containers, but this
+				// will require writing a union in place algorithm for an array container
+				// that accepts multiple different containers to union into it for
+				// efficiency.
+				container := target.Containers.Get(iKey)
+				// If target already has a bitmap container for iKey then we can reuse that,
+				// otherwise we have to allocate a new one.
+				if container == nil || container.containerType != containerBitmap {
+					buf := make([]uint64, bitmapN)
+					ob := buf[:bitmapN]
+					container = &Container{
+						bitmap:        ob,
+						n:             0,
+						containerType: containerBitmap,
+					}
+				}
+
+				// Once we've acquired a bitmap container (either by reusing the existing one
+				// or allocating a new one) then the last step is to iterate through all the
+				// other containers to see which ones have the same key, and union all of them
+				// into the target bitmap container. Only need to loop starting from i because
+				// anything previous to that has already been handled.
+				for j, jIter := range bitmapIters[i:] {
+					jKey, jContainer := jIter.iter.Value()
+
+					if iKey == jKey {
+						if jContainer.isArray() {
+							unionBitmapArrayInPlace(container, jContainer)
+						} else if jContainer.isRun() {
+							unionBitmapRunInPlace(container, jContainer)
+						} else {
+							unionBitmapBitmapInPlace(container, jContainer)
+						}
+						bitmapIters[j].handled = true
+					}
+				}
+
+				// Now that we've calculated a container is that a union of all the containers
+				// with the same key across all the bitmaps, we store it in the list of containers
+				// for the target.
+				target.Containers.Put(iKey, container)
+			}
+		}
+
+		hasNext = bitmapIters.next()
+	}
+
+	// Performing the popcount() operation with every union is wasteful because
+	// its likely the value will be invalidated by the next union operation. As
+	// a result, when we're performing all of our in-place unions we allow the value of
+	// n (container cardinality) to fall out of sync, and then at the very end we perform
+	// a "Repair" to recalculate all the container values. That way we never popcount()
+	// an entire bitmap container more than once per bulk union operation.
+	target.Containers.(*sliceContainers).repairBitmaps()
```

Comment:
Yeah I actually had it like that, but it felt a little leaky, especially since it seems like none of the functions in this package would work if containers were anything other than slice (all the functions assume a container is an array, bitmap, or RLE).

Could you give me some guidance here? My thought process is add Repair() to Container interface, then in the sliceContainer implementation it will only repair bitmaps and in the tree implementation its a no-op. Is that reasonable?

Reply:
the containers themselves are always bitmap/array/rle. The btree is for tracking which containers actually have data and where they are. This becomes a performance problem when you have potentially 10s of millions of containers in a single bitmap (this actually can happen in Pilosa due to the way we store our fragments).

now that I've looked at it a bit more, though, `repairBitmaps` doesn't actually need to be a method on `sliceContainers`, it could be a function that takes an implementation of `Containers`, and iterates through the containers calling repair on each one. e.g.

```func RepairContainers(c Containers) {
    citer := c.Iterator()
    // iterate through the containers calling bitmapRepair on each
}
```

Though if this is significantly slower, we may need separate direct implementations.

Does that make sense? I think our naming could be better, but the takeaway is that what's inside a container is always the same, but the data structure tracking the containers might be different.

Reply:
should look something like this @richardartoul

Reply:
@tgruben wouldn't this also require exposing `IsBitmap()` and `BitmapRepair()` on the `Container` type?

Reply:
right...i forgot that wasn't in the roaring package

Reply:
it can probably just be something like:

```golang
// roaring package
type Containers interface {
    ...
    Repair()
}

func (c *Container) Repair() {
    if c.isBitmap() {
        c.bitmapRepair()
    }
}
```

```golang
// btree package
func (btc *btreeContainers) Repair() {
    e, _ := btc.tree.Seek(0)
    _, c, err := e.Next()
    for err != io.EOF {
        c.Repair()
    }
    _, c, err = e.Next()
}
```

Hows that look?

Reply:
i like it.   But you need to put`e.Next()`  inside the loop